File: FarmaciasDeGuardia/Services/PDFParsing/Strategies/CuellarParser.swift
import Foundation
import PDFKit

class CuellarParser: PDFParsingStrategy {
    /// Debug flag - when true, prints detailed parsing information
    private let debug = true
    
    /// Current year being processed, incremented when January 1st is found
    private var currentYear = 2024
    
    // Lookup tables for Spanish names
    private let weekdays = [
        1: "Domingo",
        2: "Lunes",
        3: "Martes",
        4: "Miércoles",
        5: "Jueves",
        6: "Viernes",
        7: "Sábado"
    ]
    
    private let months = [
        1: "Enero",
        2: "Febrero",
        3: "Marzo",
        4: "Abril",
        5: "Mayo",
        6: "Junio",
        7: "Julio",
        8: "Agosto",
        9: "Septiembre",
        10: "Octubre",
        11: "Noviembre",
        12: "Diciembre"
    ]
    
    // Pharmacy information lookup table
    private let pharmacyInfo: [String: (name: String, address: String, phone: String)] = [
        "Av C.J. CELA": (
            name: "Farmacia Fernando Redondo",
            address: "Av. Camilo Jose Cela, 46, 40200 Cuéllar, Segovia",
            phone: "No disponible"
        ),
        "Ctra. BAHABON": (
            name: "Farmacia San Andrés",
            address: "Ctra. Bahabón, 9, 40200 Cuéllar, Segovia",
            phone: "921144794"
        ),
        "C/ RESINA": (
            name: "Farmacia Ldo. Fco. Javier Alcaraz García de la Barrera",
            address: "C. Resina, 14, 40200 Cuéllar, Segovia",
            phone: "921144812"
        ),
        "STA. MARINA": (
            name: "Farmacia Ldo. César Cabrerizo Izquierdo",
            address: "Calle Sta. Marina, 5, 40200 Cuéllar, Segovia",
            phone: "921140606"
        )
    ]
    
    /// Converts a weekday number to Spanish name
    private func weekdayName(from weekday: Int) -> String {
        return weekdays[weekday] ?? "Unknown"
    }
    
    /// Converts a month number to Spanish name
    private func monthName(from month: Int) -> String {
        return months[month] ?? "Unknown"
    }
    
    func parseSchedules(from pdfDocument: PDFDocument) -> [PharmacySchedule] {
        var schedules: [PharmacySchedule] = []
        let pageCount = pdfDocument.pageCount
        
        if debug { print("📄 Processing \(pageCount) pages...") }
        
        for pageIndex in 0..<pageCount {
            guard let page = pdfDocument.page(at: pageIndex) else {
                if debug { print("❌ Could not get page \(pageIndex) of PDF") }
                continue
            }
            
            if debug { print("\n📃 Processing page \(pageIndex + 1) of \(pageCount)") }
            
            let content = page.string ?? ""
            let lines = content.components(separatedBy: .newlines)
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            
            if debug {
                print("\n📊 Page content structure:")
                for (index, line) in lines.enumerated() {
                    print("Line \(index): '\(line)'")
                }
            }
            
            // Process the table structure for this page
            if let monthSchedules = processPageTable(lines: lines) {
                schedules.append(contentsOf: monthSchedules)
            }
        }
        
        return schedules
    }
    
    /// Represents a weekly schedule entry
    private struct WeekSchedule {
        let startDate: String  // e.g., "01-ene"
        let endDate: String    // e.g., "07-ene"
        let pharmacy: String   // e.g., "FARMACIA TORRE"
    }
    
    /// Process the table structure for a single page
    private func processPageTable(lines: [String]) -> [PharmacySchedule]? {
        var schedules: [PharmacySchedule] = []
        var pendingDates: [String]? = nil
        
        // Skip the header lines (first 2 lines)
        let dataLines = Array(lines.dropFirst(2))
        
        for line in dataLines {
            if debug { print("\n🔍 Processing line: '\(line)'") }
            
            // Skip month indicator lines and empty lines
            if line.isEmpty || (try? NSRegularExpression(pattern: "^[A-Z]{3}\\s*$").firstMatch(in: line, range: NSRange(line.startIndex..., in: line))) != nil {
                if debug { print("⏭️ Skipping empty or month indicator line") }
                continue
            }
            
            // Check if we have pending dates from a special format line
            if let dates = pendingDates {
                if debug { print("📅 Have pending dates: \(dates)") }
                // This line should be just the pharmacy name
                let pharmacy = line.trimmingCharacters(in: .whitespaces)
                if !pharmacy.isEmpty && pharmacyInfo.keys.contains(pharmacy) {
                    if debug { print("🏥 Found matching pharmacy: \(pharmacy)") }
                    processDateSet(dates: dates, pharmacy: pharmacy, into: &schedules)
                    pendingDates = nil
                    continue
                } else {
                    if debug { print("⚠️ Line doesn't match expected pharmacy format: '\(pharmacy)'") }
                }
            }
            
            // Extract dates and pharmacy from the line
            if let (dates, pharmacy) = extractDatesAndPharmacy(from: line) {
                if debug { print("📆 Extracted dates: \(dates), pharmacy: '\(pharmacy)'") }
                if pharmacy.isEmpty {
                    // This is a special format line, store the dates for the next line
                    if debug { print("⏳ Storing dates for next line") }
                    pendingDates = dates
                    continue
                }
                processDateSet(dates: dates, pharmacy: pharmacy, into: &schedules)
            }
        }
        
        return schedules
    }
    
    /// Extract dates and pharmacy from a line
    private func extractDatesAndPharmacy(from line: String) -> ([String], String)? {
        // If this is just a pharmacy name following a special format line, skip it
        if pharmacyInfo.keys.contains(line.trimmingCharacters(in: .whitespaces)) {
            if debug { print("🏥 Found standalone pharmacy line") }
            return nil
        }
        
        // Regular expression to match dates in format dd-mmm (with figure dash)
        let datePattern = #"\d{1,2}[‐-]\w{3}"#
        let regex = try? NSRegularExpression(pattern: datePattern)
        let range = NSRange(line.startIndex..., in: line)
        let matches = regex?.matches(in: line, range: range) ?? []
        
        if debug {
            print("🔍 Found \(matches.count) regular dates in line")
            // Print each character's Unicode value to help debug dash types
            print("📝 Line characters:")
            for (i, char) in line.unicodeScalars.enumerated() {
                print("   \(i): '\(char)' (Unicode: U+\(String(format:"%04X", char.value)))")
                if i > 20 { break } // Only show first few characters
            }
        }
        
        // If we found regular dates, parse them
        if !matches.isEmpty {
            var dates: [String] = []
            if debug { print("🔍 Date matches found:") }
            for match in matches {
                if let range = Range(match.range, in: line) {
                    let date = String(line[range])
                    if debug { print("   - Found date: '\(date)'") }
                    dates.append(date)
                }
            }
            
            // Get everything after the last date as the pharmacy name
            let lastDate = dates.last ?? ""
            if let pharmacyStartRange = line.range(of: lastDate)?.upperBound {
                let pharmacy = String(line[pharmacyStartRange...])
                    .trimmingCharacters(in: .whitespaces)
                
                if debug { print("📅 Regular dates: \(dates)") }
                if debug { print("🏥 Pharmacy: \(pharmacy)") }
                
                if !dates.isEmpty && !pharmacy.isEmpty {
                    return (dates, pharmacy)
                }
            }
        }
        
        // If no regular dates found, try special format (September transition)
        if line.contains("DOMINGO") || line.contains("MARTES") || line.contains("JUEVES") || line.contains("SABADO") {
            if debug { print("🎯 Found special format line") }
            if let specialTransition = parseSeptemberTransition(from: line) {
                if debug { print("✅ Parsed special dates: \(specialTransition.0)") }
                return specialTransition
            }
        }
        
        // If we get here, we couldn't parse the line
        return nil
    }
    
    /// Parse special format for September transition period and return dates with empty pharmacy
    private func parseSeptemberTransition(from line: String) -> ([String], String)? {
        if let dates = parseSeptemberTransitionDates(from: line) {
            return (dates, "")
        }
        return nil
    }
    
    /// Parse special format for September transition period and return just the dates
    private func parseSeptemberTransitionDates(from line: String) -> [String]? {
        if debug { print("\n🔄 Parsing special format dates only: '\(line)'") }
        
        // Patterns we need to handle:
        // "DOMINGO 31 DE AGOSTO Y LUNES 1 DE SEPTIEMBRE" followed by pharmacy
        // "MARTES 2 Y MIERCOLES 3 de SEPTIEMBRE" followed by pharmacy
        // "JUEVES 4 Y VIERNES 5 de SEPTIEMBRE" followed by pharmacy
        // "SABADO 6 Y DOMINGO 7 de Septiembre" followed by pharmacy
        
        // Handle different September transition patterns
        // Like "31 DE AGOSTO" or just "2" in "MARTES 2"
        let datePattern = #"(?:(\d+)\s+DE\s+(AGOSTO|SEPTIEMBRE|Septiembre)|(?:LUNES|MARTES|MIERCOLES|JUEVES|VIERNES|SABADO|DOMINGO)\s+(\d+)(?:\s+[Dd][Ee]\s+(?:SEPTIEMBRE|Septiembre))?)"#
        let regex = try? NSRegularExpression(pattern: datePattern)
        
        guard let matches = regex?.matches(in: line, range: NSRange(line.startIndex..., in: line)) else {
            if debug { print("❌ No matches found in line") }
            return nil
        }
        
        if debug { print("✅ Found \(matches.count) date matches") }
        
        var dates: [String] = []
        for match in matches {
            var day: String?
            var month: String?
            
            // Try first pattern: "31 DE AGOSTO"
            if let dayRange = Range(match.range(at: 1), in: line),
               let monthRange = Range(match.range(at: 2), in: line) {
                day = String(line[dayRange])
                month = String(line[monthRange].lowercased())
            }
            // Try second pattern: "MARTES 2"
            else if let dayRange = Range(match.range(at: 3), in: line) {
                day = String(line[dayRange])
                month = line.contains("SEPTIEMBRE") || line.contains("Septiembre") ? "septiembre" : "agosto"
            }
            
            if let day = day, let month = month {
                let monthAbbr = String(month.prefix(3))
                dates.append("\(String(format: "%02d", Int(day) ?? 0))-\(monthAbbr)")
            }
        }
        
        if debug { print("📅 Parsed dates: \(dates)") }
        return dates.isEmpty ? nil : dates
    }
    
    /// Converts a Spanish abbreviated month to a month number
    private func monthNumber(from abbreviation: String) -> Int? {
        let months = [
            "ene": 1, "feb": 2, "mar": 3, "abr": 4, "may": 5, "jun": 6,
            "jul": 7, "ago": 8, "sep": 9, "oct": 10, "nov": 11, "dic": 12
        ]
        return months[abbreviation.lowercased()]
    }
    
    /// Converts a date string like "01-ene" to a DutyDate
    private func parseDutyDate(_ dateString: String, year: Int) -> DutyDate? {
        let pattern = #"(\d{1,2})[‐-](\w{3})"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: dateString, range: NSRange(dateString.startIndex..., in: dateString)),
              let dayRange = Range(match.range(at: 1), in: dateString),
              let monthRange = Range(match.range(at: 2), in: dateString),
              let day = Int(String(dateString[dayRange])),
              let month = monthNumber(from: String(dateString[monthRange])) else {
            return nil
        }
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.locale = Locale(identifier: "es_ES")
        let dateString = String(format: "%d-%02d-%02d", year, month, day)
        
        guard let date = dateFormatter.date(from: dateString) else {
            return nil
        }
        
        let calendar = Calendar.current
        let weekday = calendar.component(.weekday, from: date)
        
        return DutyDate(
            dayOfWeek: weekdayName(from: weekday),
            day: day,
            month: monthName(from: month),
            year: year
        )
    }
    
    /// Expand a date range into individual DutyDate objects
    private func expandWeekDates(start: String, end: String, year: Int) -> [DutyDate]? {
        // Convert start and end strings to dates
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "dd-MMM-yyyy"
        dateFormatter.locale = Locale(identifier: "es_ES")
        
        // Append year to the date strings
        let startWithYear = start + "-\(year)"
        let endWithYear = end + "-\(year)"
        
        guard let startDate = dateFormatter.date(from: startWithYear),
              let endDate = dateFormatter.date(from: endWithYear) else {
            return nil
        }
        
        var dates: [DutyDate] = []
        let calendar = Calendar.current
        let daysBetween = calendar.dateComponents([.day], from: startDate, to: endDate).day ?? 0
        
        for dayOffset in 0...daysBetween {
            guard let currentDate = calendar.date(byAdding: .day, value: dayOffset, to: startDate) else {
                continue
            }
            
            let components = calendar.dateComponents([.year, .month, .day, .weekday], from: currentDate)
            guard let month = components.month,
                  let day = components.day,
                  let weekday = components.weekday else {
                continue
            }
            
            let dutyDate = DutyDate(
                dayOfWeek: weekdayName(from: weekday),
                day: day,
                month: monthName(from: month),
                year: year
            )
            
            dates.append(dutyDate)
        }
        
        return dates
    }
    
    
    /// Process a set of dates with a pharmacy
    private func processDateSet(dates: [String], pharmacy: String, into schedules: inout [PharmacySchedule]) {
        if debug { print("\n📋 Processing date set:") }
        if debug { print("📅 Dates: \(dates)") }
        if debug { print("🏥 Pharmacy: \(pharmacy)") }
        if debug { print("📆 Current year: \(currentYear)") }
        
        // Create schedules for each date in the set
        for date in dates {
            // If this is January 1st, increment the year
            let pattern = #"01[‐-]ene"#
            if let regex = try? NSRegularExpression(pattern: pattern),
               regex.firstMatch(in: date, range: NSRange(date.startIndex..., in: date)) != nil {
                currentYear += 1
                if debug { print("🎊 New year detected! Now processing year \(currentYear)") }
            }

            if debug { print("📆 Processing date: \(date) (year: \(currentYear))") }
            
            if let dutyDate = parseDutyDate(date, year: currentYear) {
                let info = pharmacyInfo[pharmacy] ?? (
                    name: pharmacy,
                    address: "Dirección no disponible",
                    phone: "No disponible"
                )
                
                let pharmacyInstance = Pharmacy(
                    name: info.name,
                    address: info.address,
                    phone: info.phone,
                    additionalInfo: nil
                )
                
                let schedule = PharmacySchedule(
                    date: dutyDate,
                    dayShiftPharmacies: [pharmacyInstance],
                    nightShiftPharmacies: [] // Cuéllar doesn't use separate night shifts
                )
                
                schedules.append(schedule)
            }
        }
    }
}

File: android/app/src/main/java/com/example/farmaciasdeguardiaensegovia/services/pdfparsing/strategies/CuellarParser.kt
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.example.farmaciasdeguardiaensegovia.services.pdfparsing.strategies

import com.example.farmaciasdeguardiaensegovia.data.DutyDate
import com.example.farmaciasdeguardiaensegovia.data.DutyTimeSpan
import com.example.farmaciasdeguardiaensegovia.data.Pharmacy
import com.example.farmaciasdeguardiaensegovia.data.PharmacySchedule
import com.example.farmaciasdeguardiaensegovia.services.DebugConfig
import com.example.farmaciasdeguardiaensegovia.services.pdfparsing.PDFParsingStrategy
import com.example.farmaciasdeguardiaensegovia.services.pdfparsing.PDFParsingUtils
import com.example.farmaciasdeguardiaensegovia.services.pdfparsing.strategies.ElEspinarParser.PharmacyInfo
import com.itextpdf.kernel.pdf.PdfDocument
import com.itextpdf.kernel.pdf.PdfReader
import com.itextpdf.kernel.pdf.canvas.parser.PdfTextExtractor
import java.io.File

/**
 * Parser implementation for Cuéllar pharmacy schedules.
 * Android equivalent of iOS CuellarParser with enhanced special case handling.
 * 
 * Handles both regular format (dd-mmm) and special transition format 
 * (e.g., "DOMINGO 31 DE AGOSTO Y LUNES 1 DE SEPTIEMBRE").
 */
class CuellarParser : PDFParsingStrategy {
    
    /** Current year being processed, incremented when January 1st is found */
    private var currentYear = 2024
    
    override fun getStrategyName(): String = "CuellarParser"
    
    override fun parseSchedules(pdfFile: File): List<PharmacySchedule> {
        val allSchedules = mutableListOf<PharmacySchedule>()
        
        DebugConfig.debugPrint("\n=== Cuéllar Pharmacy Schedules ===")
        
        // Open PDF once and reuse across all pages
        val reader = PdfReader(pdfFile)
        val pdfDoc = PdfDocument(reader)
        
        return try {
            val pageCount = pdfDoc.numberOfPages
            DebugConfig.debugPrint("📄 Processing $pageCount pages of Cuéllar PDF...")
            
            // Process each page
            for (pageIndex in 1..pageCount) { // iText uses 1-based indexing
                DebugConfig.debugPrint("\n📃 Processing page $pageIndex of $pageCount")
                
                val content = PdfTextExtractor.getTextFromPage(pdfDoc.getPage(pageIndex))
                val lines = content.split('\n')
                    .map { it.trim() }
                    .filter { it.isNotEmpty() }
                
                if (lines.isNotEmpty()) {
                    DebugConfig.debugPrint("\n📊 Page content structure:")
                    lines.forEachIndexed { index, line ->
                        DebugConfig.debugPrint("Line $index: '$line'")
                    }
                }
                
                // Process the table structure for this page
                val pageSchedules = processPageTable(lines)
                allSchedules.addAll(pageSchedules)
            }
            
            // Sort schedules by date efficiently
            val sortedSchedules = allSchedules.sortedWith(dateComparator)
            
            DebugConfig.debugPrint("✅ Successfully parsed ${sortedSchedules.size} schedules for Cuéllar")
            sortedSchedules
            
        } catch (e: Exception) {
            DebugConfig.debugError("❌ Error parsing Cuéllar PDF: ${e.message}", e)
            emptyList()
        } finally {
            pdfDoc.close()
        }
    }

    private fun processPageTable(lines: List<String>): List<PharmacySchedule> {
        val (schedules, _, _) = lines.fold(Triple(emptyList<PharmacySchedule>(), null as String?, emptyList<String>())) { (acc, pharmacyKey, dates), line ->
            DebugConfig.debugPrint("\n🔍 Processing line: '$line'")

            val (parsedDates, parsedPharmacy) = when {
                hasDates(line) -> processCompositeLine(line)
                hasPharmacy(line) -> {
                    val maybePharmacy = extractPharmacyFromLine(line)
                    Pair(dates, maybePharmacy)
                }
                else -> {
                    DebugConfig.debugPrint("⏭️ Skipping unsupported line: [$line]")
                    Pair(dates, pharmacyKey)
                }
            }

            if(parsedPharmacy != null && parsedDates.isNotEmpty()) {
                Triple(acc + processDateSet(parsedDates, parsedPharmacy), null, emptyList())
            } else Triple(acc, parsedPharmacy, parsedDates)
        }

        return schedules
    }

    private fun hasDates(line: String): Boolean = REGULAR_DATE_REGEX.containsMatchIn(line)

    private fun hasPharmacy(line: String): Boolean =
        normalizeWhitespace(line).let { normalizedLine ->
            PHARMACY_INFO.keys.find { pharmacyKey ->
                normalizedLine.contains(pharmacyKey, ignoreCase = true)
            } != null
        }

    private fun processCompositeLine(line: String): Pair<List<String>, String?> {
        val dates = REGULAR_DATE_REGEX.findAll(line).map { it.value }.toList()

        val maybePharmacy = extractPharmacyFromLine(line)

        if(maybePharmacy == null) {
            DebugConfig.debugPrint("Composite line had no pharmacy information: [$line]")
            DebugConfig.debugPrint("Will try to find pharmacy information next line")
        }
        return Pair(dates, maybePharmacy)
    }

    private fun extractPharmacyFromLine(line: String): String? =
        normalizeWhitespace(line).let { normalizedLine ->
            PHARMACY_INFO.keys.find { pharmacyKey ->
                normalizedLine.contains(pharmacyKey, ignoreCase = true)
            }
    }
    
    /**
     * Check if line contains special transition format (e.g., "DE AGOSTO", "DE SEPTIEMBRE")
     */
    private fun containsSpecialTransition(line: String): Boolean {
        return line.contains("DE AGOSTO", ignoreCase = true) || 
               line.contains("DE SEPTIEMBRE", ignoreCase = true) ||
               line.contains("de Septiembre")
    }
    
    /**
     * Check if line contains regular date format (dd-mmm)
     */
    private fun containsRegularDates(line: String): Boolean {
        return REGULAR_DATE_REGEX.containsMatchIn(line)
    }
    
    /**
     * Check if line is a month indicator (e.g., "ENE", "FEB", "MAR")
     */
    private fun isMonthIndicatorLine(line: String): Boolean {
        return MONTH_INDICATOR_REGEX.matches(line)
    }
    
    /**
     * Parse special transition format lines like:
     * "DOMINGO 31 DE AGOSTO Y LUNES 1 DE SEPTIEMBRE  STA. MARINA"
     */
    private fun parseSpecialTransitionLine(line: String): Pair<List<String>, String>? {
        DebugConfig.debugPrint("\n🔄 Parsing special format line: '$line'")
        
        val dates = mutableListOf<String>()
        val dateMatches = SPECIAL_TRANSITION_REGEX.findAll(line)
        
        DebugConfig.debugPrint("✅ Found ${dateMatches.count()} special date matches")
        
        for (match in dateMatches) {
            val day = match.groupValues[2].toIntOrNull() ?: continue
            val monthAbbr = when (match.groupValues[3].uppercase()) {
                "AGOSTO" -> "ago"
                "SEPTIEMBRE", "Septiembre" -> "sep"
                else -> continue
            }
            val formattedDate = String.format("%02d-%s", day, monthAbbr)
            dates.add(formattedDate)
            DebugConfig.debugPrint("   - Converted '${match.value}' to '$formattedDate'")
        }
        
        // Extract pharmacy name from end of line
        val pharmacy = extractPharmacyFromEndOfLine(line)
        
        DebugConfig.debugPrint("📅 Special format dates: $dates")
        DebugConfig.debugPrint("🏥 Pharmacy: '$pharmacy'")
        
        return if (dates.isNotEmpty() && pharmacy.isNotEmpty()) {
            Pair(dates, pharmacy)
        } else {
            DebugConfig.debugPrint("❌ Failed to extract dates or pharmacy from special format line")
            null
        }
    }
    
    /**
     * Parse regular format lines like:
     * "30-dic  31-dic  01-ene  02-ene  03-ene  04-ene  05-ene  Av C.J. CELA"
     */
    private fun parseRegularDateLine(line: String): Pair<List<String>, String>? {
        DebugConfig.debugPrint("\n📋 Parsing regular format line: '$line'")
        
        val dates = mutableListOf<String>()
        val dateMatches = REGULAR_DATE_REGEX.findAll(line)
        
        DebugConfig.debugPrint("✅ Found ${dateMatches.count()} regular date matches")
        
        for (match in dateMatches) {
            dates.add(match.value)
            DebugConfig.debugPrint("   - Found date: '${match.value}'")
        }
        
        // Extract pharmacy name from end of line (after the last date)
        val pharmacy = extractPharmacyFromEndOfLine(line)
        
        DebugConfig.debugPrint("📅 Regular dates: $dates")
        DebugConfig.debugPrint("🏥 Pharmacy: '$pharmacy'")
        
        return if (dates.isNotEmpty() && pharmacy.isNotEmpty()) {
            Pair(dates, pharmacy)
        } else {
            DebugConfig.debugPrint("❌ Failed to extract dates or pharmacy from regular format line")
            null
        }
    }
    
    /**
     * Normalize whitespace by replacing all types of whitespace characters with single regular spaces
     * This handles NBSP (non-breaking space), tabs, multiple spaces, etc.
     */
    private fun normalizeWhitespace(text: String): String {
        // Replace all Unicode whitespace characters with regular spaces, then collapse multiple spaces
        return text.replace(WHITESPACE_REGEX, " ").trim()
    }
    
    /**
     * Extract pharmacy name from the end of a line
     */
    private fun extractPharmacyFromEndOfLine(line: String): String {
        DebugConfig.debugPrint("🔍 Extracting pharmacy from line: '$line'")
        
        // Normalize whitespace in the input line
        val normalizedLine = normalizeWhitespace(line)
        DebugConfig.debugPrint("🔧 Normalized line: '$normalizedLine'")
        
        // Find the pharmacy name that appears in the line (with normalized whitespace comparison)
        val foundPharmacy = PHARMACY_INFO.keys.find { pharmacy ->
            val normalizedPharmacy = normalizeWhitespace(pharmacy.trim())
            val matches = normalizedLine.contains(normalizedPharmacy, ignoreCase = true)
            DebugConfig.debugPrint("🔍 Checking '$normalizedPharmacy' against normalized line: $matches")
            matches
        }?.trim()
        
        DebugConfig.debugPrint("🔍 Found pharmacy match: '$foundPharmacy'")
        return foundPharmacy ?: ""
    }
    
    /**
     * Process a set of dates with a pharmacy to create schedules
     */
    private fun processDateSet(dates: List<String>, pharmacyKey: String): List<PharmacySchedule> {
        DebugConfig.debugPrint("📋 Processing date set:")
        DebugConfig.debugPrint("📅 Dates: $dates")
        DebugConfig.debugPrint("🏠 Pharmacy: $pharmacyKey")
        DebugConfig.debugPrint("📆 Current year: $currentYear")

        return dates.fold(emptyList<PharmacySchedule>()) { acc, date ->
            if (date.matches(Regex("01[‐-]ene"))) {
                currentYear++
                DebugConfig.debugPrint("🎊 New year detected! Now processing year $currentYear")
            }

            DebugConfig.debugPrint("📆 Processing date: $date (year: $currentYear)")
            val dutyDate = parseDutyDate(date, currentYear)

            dutyDate?.let { dutyDate ->
                val pharmacyInfo = PHARMACY_INFO[pharmacyKey] ?: PharmacyInfo(
                    name = "Farmacia $pharmacyKey",
                    address = "Dirección no disponible",
                    phone = "No disponible"
                )

                val pharmacyInstance = Pharmacy(
                    name = pharmacyInfo.name,
                    address = pharmacyInfo.address,
                    phone = pharmacyInfo.phone,
                    additionalInfo = null
                )

                DebugConfig.debugPrint("💊 Adding schedule for ${pharmacyInstance.name} on ${dutyDate.day}-${dutyDate.month}-${dutyDate.year ?: PDFParsingUtils.getCurrentYear()}")

                acc + PharmacySchedule(
                    date = dutyDate,
                    shifts = mapOf(
                        DutyTimeSpan.FullDay to listOf(pharmacyInstance)
                    )
                )
            } ?: acc.also {
                DebugConfig.debugPrint("⚠️ Could not parse duty date for: $date")
            }
        }
    }
    
    /**
     * Parse a date string like "01-ene" to a DutyDate
     */
    private fun parseDutyDate(dateString: String, year: Int): DutyDate? {
        val match = REGULAR_DATE_REGEX.matchEntire(dateString) ?: return null
        
        val dayStr = match.groupValues[1]
        val monthStr = match.groupValues[2]
        
        val day = dayStr.toIntOrNull() ?: return null
        val month = PDFParsingUtils.monthAbbrToNumber(monthStr) ?: return null
        
        return DutyDate(
            dayOfWeek = PDFParsingUtils.getDayOfWeek(day, month, year),
            day = day,
            month = PDFParsingUtils.getMonthName(month),
            year = year
        )
    }
    
    /**
     * Data class for pharmacy information
     */
    private data class PharmacyInfo(
        val name: String,
        val address: String,
        val phone: String
    )
    
    companion object {
        // Pre-compiled regex patterns for performance
        private val REGULAR_DATE_REGEX by lazy { 
            Regex("""(\d{1,2})[‐-](\w{3})""") 
        }
        
        private val SPECIAL_TRANSITION_REGEX by lazy { 
            Regex("""(?:DOMINGO|LUNES|MARTES|MIERCOLES|JUEVES|VIERNES|SABADO)\s+(\d+)\s+DE\s+(AGOSTO|SEPTIEMBRE|Septiembre)""", RegexOption.IGNORE_CASE) 
        }
        
        private val MONTH_INDICATOR_REGEX by lazy { 
            Regex("""^[A-Z]{3}\s*$""") 
        }
        
        // Regex to match all types of whitespace characters (NBSP, regular space, tab, etc.)
        private val WHITESPACE_REGEX by lazy {
            Regex("""[\s\u00A0\u2000-\u200A\u2028\u2029\u202F\u205F\u3000]+""")
        }
        
        // Pharmacy information lookup table (matching iOS implementation)
        private val PHARMACY_INFO = mapOf(
            "Av C.J. CELA" to PharmacyInfo(
                name = "Farmacia Fernando Redondo",
                address = "Av. Camilo Jose Cela, 46, 40200 Cuéllar, Segovia",
                phone = "No disponible"
            ),
            "Ctra. BAHABON" to PharmacyInfo(
                name = "Farmacia San Andrés",
                address = "Ctra. Bahabón, 9, 40200 Cuéllar, Segovia",
                phone = "921144794"
            ),
            "C/ RESINA" to PharmacyInfo(
                name = "Farmacia Ldo. Fco. Javier Alcaraz García de la Barrera",
                address = "C. Resina, 14, 40200 Cuéllar, Segovia",
                phone = "921144812"
            ),
            "STA. MARINA" to PharmacyInfo(
                name = "Farmacia Ldo. César Cabrerizo Izquierdo",
                address = "Calle Sta. Marina, 5, 40200 Cuéllar, Segovia",
                phone = "921140606"
            )
        )
        
        // Cached date comparator to avoid lambda creation overhead
        private val dateComparator = Comparator<PharmacySchedule> { first, second ->
            compareSchedulesByDate(first, second)
        }
        
        /**
         * Compare two pharmacy schedules by date for sorting
         */
        private fun compareSchedulesByDate(first: PharmacySchedule, second: PharmacySchedule): Int {
            val currentYear = PDFParsingUtils.getCurrentYear()
            
            // Extract year, month, day from dates
            val firstYear = first.date.year ?: currentYear
            val secondYear = second.date.year ?: currentYear
            
            if (firstYear != secondYear) {
                return firstYear.compareTo(secondYear)
            }
            
            val firstMonth = PDFParsingUtils.monthToNumber(first.date.month) ?: 0
            val secondMonth = PDFParsingUtils.monthToNumber(second.date.month) ?: 0
            
            if (firstMonth != secondMonth) {
                return firstMonth.compareTo(secondMonth)
            }
            
            val firstDay = first.date.day
            val secondDay = second.date.day
            
            return firstDay.compareTo(secondDay)
        }
    }
}
